config/stack.go

```diff
@@ -40,12 +47,251 @@ type Unit struct {
 	Values *cty.Value `hcl:"values,attr"`
 }
 
+// Stack represents the stack block in the configuration.
+type Stack struct {
+	Name   string `hcl:",label"`
+	Source string `hcl:"source,attr"`
+	Path   string `hcl:"path,attr"`
+}
+
+// GenerateStacks generates the stack files.
+func GenerateStacks(ctx context.Context, opts *options.TerragruntOptions) error {
+	processedFiles := make(map[string]bool)
+	// initial files setting as stack file
+	foundFiles := []string{opts.TerragruntStackConfigPath}
+
+	for {
+		// check if we have already processed the files
+		processedNewFiles := false
+
+		for _, file := range foundFiles {
+			if processedFiles[file] {
+				continue
+			}
+
+			processedNewFiles = true
+			processedFiles[file] = true
+
+			if err := processStackFile(ctx, opts, file); err != nil {
+				return errors.Errorf("Failed to process stack file %s %v", file, err)
+			}
+		}
+
+		if !processedNewFiles {
+			break
+		}
+
+		newFiles, err := listStackFiles(opts, opts.WorkingDir)
+
+		if err != nil {
+			return errors.Errorf("Failed to list stack files %v", err)
+		}
+
+		foundFiles = newFiles
+	}
+
+	return nil
+}
+
+// StackOutput generates the output from the stack files.
+func StackOutput(ctx context.Context, opts *options.TerragruntOptions) (map[string]map[string]cty.Value, error) {
+	opts.Logger.Debugf("Generating output from %s", opts.TerragruntStackConfigPath)
+	opts.TerragruntStackConfigPath = filepath.Join(opts.WorkingDir, defaultStackFile)
+	stackTargetDir := filepath.Join(opts.WorkingDir, stackDir)
+	stackFiles, err := listStackFiles(opts, stackTargetDir)
+
+	if err != nil {
+		return nil, errors.Errorf("Failed to list stack files in %s %v", stackTargetDir, err)
+	}
+
+	unitOutputs := make(map[string]map[string]cty.Value)
+
+	// add default stack file
+	stackFiles = append(stackFiles, opts.TerragruntStackConfigPath)
```

Comment:
Should we check if the file exists before we do this? I wonder if we should be agnostic to where a user runs the command.

Reply:
now, it fails later with error:
```
20:39:38.369 ERROR  Failed to process stack file ./terragrunt.stack.hcl Failed to read stack file ./terragrunt.stack.hcl in . open ./terragrunt.stack.hcl: no such file or directory
20:39:38.371 ERROR  Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```